src/freenet/node/BandwidthManager.java

```diff
@@ -0,0 +1,109 @@
+package freenet.node;
+
+import freenet.clients.http.wizardsteps.BandwidthLimit;
+import freenet.config.InvalidConfigValueException;
+import freenet.l10n.NodeL10n;
+import freenet.node.useralerts.UpgradeConnectionSpeedUserAlert;
+import freenet.pluginmanager.FredPluginBandwidthIndicator;
+import freenet.support.Logger;
+
+import static java.util.concurrent.TimeUnit.*;
+
+public class BandwidthManager {
+
+    private static final long DELAY_HOURS = 24;
+
+    private int lastOfferedInputBandwidth;
+    private int lastOfferedOutputBandwidth;
+
+    private final Node node;
+
+    BandwidthManager(Node node) {
+        this.node = node;
+    }
+
+    public void start() {
+        // TODO: move to "on upgrade"?
+        /* offer upgrade of the connection speed on upgrade, if auto-detected
+         * speed is much higher than the set speed, or even better: if the
+         * detected speed increased significantly since the last offer. */
+        node.ticker.queueTimedJob(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    FredPluginBandwidthIndicator bandwidthIndicator = node.ipDetector.getBandwidthIndicator();
+                    if (!node.config.get("node").getBoolean("connectionSpeedDetection") ||
+                            bandwidthIndicator == null) {
+                        return;
+                    }
+
+                    int detectedInputBandwidth = bandwidthIndicator.getDownstreamMaxBitRate() / 8;
+                    int detectedOutputBandwidth = bandwidthIndicator.getUpstramMaxBitRate() / 8;
+
+                    int currentInputBandwidth = node.config.get("node").getInt("inputBandwidthLimit");
+                    int currentOutputBandwidth = node.config.get("node").getInt("outputBandwidthLimit");
+
+                    if (detectedInputBandwidth > currentInputBandwidth * 3 &&
+                            detectedInputBandwidth > lastOfferedInputBandwidth * 3 ||
+                        detectedOutputBandwidth > currentOutputBandwidth * 3 &&
+                                detectedOutputBandwidth > lastOfferedOutputBandwidth * 3) {
```

Comment:
Speed check happens once a day, yes. But the alert will appear if the speed is significantly greater than previously offered one.

Reply:
that sounds good; I had missed that — thank you!